beacon-chain/db/kv/blob.go

```diff
@@ -45,77 +54,73 @@ func (rk blobRotatingKey) BlockRoot() []byte {
+		var existing []byte
+		sc := &ethpb.BlobSidecars{}
 		bkt := tx.Bucket(blobsBucket)
 		c := bkt.Cursor()
-		var replacingKey blobRotatingKey
-		for k, _ := c.Seek(rotatingBufferPrefix); bytes.HasPrefix(k, rotatingBufferPrefix); k, _ = c.Next() {
-			if len(k) != 0 {
-				replacingKey = k
-				break
+		for k, v := c.Seek(prefix); bytes.HasPrefix(k, prefix); k, v = c.Next() {
+			key := blobRotatingKey(k)
+			ks := key.Slot()
+			if ks < first.Slot {
+				// Mark older blobs at the same position of the ring buffer for deletion.
+				prune = append(prune, key)
+				continue
 			}
-		}
-		var existing []byte
-		sc := &ethpb.BlobSidecars{}
-		// If there is no element stored at blob.slot % MAX_SLOTS_TO_PERSIST_BLOBS, then we simply
-		// store the blob by key and exit early.
-		if len(replacingKey) != 0 {
-			oldSlot := replacingKey.Slot()
-			oldEpoch := slots.ToEpoch(oldSlot)
-			// The blob we are replacing is too old, so we delete it.
-			if slots.ToEpoch(scs[0].Slot) >= oldEpoch.Add(uint64(params.BeaconNetworkConfig().MinEpochsForBlobsSidecarsRequest)) {
-				if err := bkt.Delete(replacingKey); err != nil {
-					log.WithError(err).Warnf("Could not delete blob with key %#x", replacingKey)
-					return err
-				}
-			} else {
-				// Otherwise, we need to merge the new blob with the old blob.
-				existing = bkt.Get(replacingKey)
-				if err := decode(ctx, existing, sc); err != nil {
+			if ks > first.Slot {
+				// We shouldn't be overwriting newer blobs with older blobs. Something is wrong.
+				return errNewerBlobExists
+			}
+			// The slot isn't older or newer, so it must be equal.
+			// If the roots match, then we want to merge the new sidecars with the existing data.
+			if bytes.Equal(first.BlockRoot, key.BlockRoot()) {
+				existing = v
+				if err := decode(ctx, v, sc); err != nil {
 					return err
 				}
 			}
+			// If the slot is equal but the roots don't match, leave the existing key alone and allow the sidecar
+			// to be written to the new key with the same prefix. In this case sc will be empty, so it will just
+			// contain the incoming sidecars when we write it.
 		}
-
 		sc.Sidecars = append(sc.Sidecars, scs...)
 		sortSidecars(sc.Sidecars)
 		var err error
 		sc.Sidecars, err = validUniqueSidecars(sc.Sidecars)
 		if err != nil {
 			return err
 		}
-		updated, err := encode(ctx, sc)
+		encoded, err := encode(ctx, sc)
 		if err != nil {
 			return err
 		}
-
 		// don't write if the merged result is the same as before
-		if len(existing) == len(updated) && bytes.Equal(existing, updated) {
+		if len(existing) == len(encoded) && bytes.Equal(existing, encoded) {
 			return nil
 		}
-
-		return bkt.Put(newKey, updated)
+		// Only prune if we're actually going through with the update.
+		for _, k := range prune {
+			if err := bkt.Delete(k); err != nil {
+				// note: attempting to delete a key that does not exist should not return an error.
+				log.WithError(err).Warnf("Could not delete blob key %#x.", k)
+			}
+		}
```

Comment:
Should we `Allow` before `validUniqueSidecars`? I'm inclined to say this is better, but if `validUniqueSidecars` or `encode` error out, we will miss the chance to prune.

Reply:
I like not touching the db at all when the data we were asked to write isn't writable.